bugbug/bugzilla.py

```diff
@@ -93,50 +94,84 @@ def bughandler(bug):
+            a_value = {
+                a_field: a_value[a_field]
+                for a_field in list(a_value.keys())
+                if a_field in required_fields
+            }
+            array_of_dict_values[a_ind] = a_value
+        return array_of_dict_values
+
+    def cleanup_fields_in_batch_of_bugs(batch_of_bugs):
+        # Updation happens in place
+        for a_bug_info_key in batch_of_bugs.keys():
+            a_bug_info = batch_of_bugs[a_bug_info_key]
+            current_comments_array = a_bug_info["comments"]
+            current_comments_array = filter_keys(
+                array_of_dict_values=current_comments_array,
+                required_fields=COMMENT_INCLUDE_FIELDS,
+            )
+            a_bug_info["comments"] = current_comments_array
+        return batch_of_bugs
 
     def attachmenthandler(bug, bug_id):
         bug_id = int(bug_id)
-
         if bug_id not in new_bugs:
             new_bugs[bug_id] = dict()
-
         new_bugs[bug_id]["attachments"] = bug
 
-    def historyhandler(bug):
-        bug_id = int(bug["id"])
-
-        if bug_id not in new_bugs:
-            new_bugs[bug_id] = dict()
-
-        new_bugs[bug_id]["history"] = bug["history"]
-
-    Bugzilla(
-        ids_or_query,
-        bughandler=bughandler,
-        commenthandler=commenthandler,
-        comment_include_fields=COMMENT_INCLUDE_FIELDS,
-        attachmenthandler=attachmenthandler,
-        attachment_include_fields=ATTACHMENT_INCLUDE_FIELDS,
-        historyhandler=historyhandler,
-    ).get_data().wait()
+    bug_ids = sorted(set(bug_ids))
+
+    new_bugs = dict()
+
+    batch_size = Bugzilla.BUGZILLA_CHUNK_SIZE
+
+    # Getting _default, history and comments information using REST API
+    # as specified
+
+    for i in range(0, len(bug_ids), batch_size):
+        batch = bug_ids[i : i + batch_size]
+        batch_of_ids = ",".join(map(str, batch))
+
+        # "include_fields": "_default,history,comments,attachments",
+        # Attachments data size is heavy, so handling them separately
+        params_for_custom_fields = {
+            "id": batch_of_ids,
+            "include_fields": "_default,history,comments",
+        }
+        response = utils.get_session("bugzilla").get(
+            "https://bugzilla.mozilla.org/rest/bug", params=params_for_custom_fields
+        )
+        response.raise_for_status()
+
+        batch_of_bugs_info = response.json()
+        batch_of_bugs_info = {
+            int(a_bug["id"]): a_bug for a_bug in batch_of_bugs_info["bugs"]
+        }
+        batch_of_bugs_info = cleanup_fields_in_batch_of_bugs(batch_of_bugs_info)
+        new_bugs.update(batch_of_bugs_info)
+
+        Bugzilla(
+            batch,
+            bughandler=None,
+            commenthandler=None,
+            comment_include_fields=None,
+            attachmenthandler=attachmenthandler,
+            attachment_include_fields=ATTACHMENT_INCLUDE_FIELDS,
+            historyhandler=None,
+        ).get_data().wait()
```

Comment:
This means we'll query the bugs/history/comment first, and then query the attachments.
Without the patch, we were querying all of them in parallel.

Reply:
we'll query the bugs/history/comment first, and then query the attachments. - yes.

"Without the patch, we were querying all of them in parallel." -  If the handlers are present, then the bug information is retrieved, followed by history info, comment info, and then attachment info in order. 
I meant  1) retrieve bug info, 2) retrieve history info 3) retrieve comment info  4) retrieve attachment info are done Sequentially one after another.

Reply:
I don't think that's the case, we were retrieving 1, 2, 3, and 4 in parallel. Why do you say they were retrieved in sequence?

Reply:
Bugzilla(..).get_data().wait() - call is propagated as below:

https://github.com/mozilla/libmozdata/blob/master/libmozdata/bugzilla.py#L152-L169

In this function we can find sequential retrieval of

1. Bug Information - https://github.com/mozilla/libmozdata/blob/master/libmozdata/bugzilla.py#L159
which calls
https://github.com/mozilla/libmozdata/blob/master/libmozdata/bugzilla.py#L556

2. History Information ->           https://github.com/mozilla/libmozdata/blob/master/libmozdata/bugzilla.py#L161
3. Comments -> https://github.com/mozilla/libmozdata/blob/master/libmozdata/bugzilla.py#L163
4. Attachments -> https://github.com/mozilla/libmozdata/blob/master/libmozdata/bugzilla.py#L165

Please let me know if my understanding is incorrect.

Reply:
Those functions are triggering requests, but not waiting for them to finish (they add futures to `self.bugs_results`, `self.history_results`, and so on). Then in `wait` all of the futures are awaited.

Reply:
Yes ,  Futuresession is used. https://github.com/mozilla/libmozdata/blob/b08f2cb144e1db97ce9113160695b6477df28b45/libmozdata/connection.py#L73 .

Thanks for clarification.
The retrieval is being done asynchrously + parallely.